runtime/src/accounts_db.rs

```diff
@@ -649,6 +649,21 @@ impl AccountsDB {
             }
         }
 
+        for account_infos in purges.values() {
+            let mut no_delete = false;
+            for (_slot_id, account_info) in account_infos {
+                if *store_counts.get(&account_info.store_id).unwrap() != 0 {
+                    no_delete = true;
+                    break;
+                }
+            }
+            if no_delete {
+                for (_slot_id, account_info) in account_infos {
```

Comment:
I carefully re-read the `purge_zero_lamport_accounts`. And I think this new logic works as intended. Very nice and thank you so much! It's sad for me not to notice nor write this... ;)

[My previous perception](https://github.com/solana-labs/solana/pull/8176#discussion_r376724915) was wrong... Especially the DAG thing. I wonder why I was urged to think so. Maybe, too tired...

And this makes `purge_zero_lamport_accounts` yet more conservative. So, regardless a hostile or normal situation, I think accumulated zero lamport accounts impose us no-op wasted CPU time by bunch of loops for each `snapshot_slot_interval` to some extent.

Especially, this loop exhibits some contaminating/cascading behavior, easily approaching to the worst cost of `O(N)` where N is the total number of all updates to every accounts which became 0 lamport in the past at least once. That's because making `no_delete` be `true` for vast of unrelated `pubkey`s can easily be done by triggering the avalanche counting just with a single zero account with many updates.

In near future, I think we'd need some compaction mechanism as implied by #8168...